logstash-core/src/main/java/org/logstash/config/ir/compiler/DatasetCompiler.java

```diff
@@ -378,14 +383,14 @@ private static MethodLevelSyntaxElement callFilterFlush(final ClassFields fields
         final boolean shutdownOnly) {
         final MethodLevelSyntaxElement condition;
         final ValueSyntaxElement flushArgs;
-        final ValueSyntaxElement flushFinal = fields.add(flushOpts(true));
+        final ValueSyntaxElement flushFinal = fields.add("flushFinalFlag", flushOpts(true));
         if (shutdownOnly) {
             condition = SyntaxFactory.and(FLUSH_ARG, SHUTDOWN_ARG);
             flushArgs = flushFinal;
         } else {
             condition = FLUSH_ARG;
             flushArgs = SyntaxFactory.ternary(
-                SHUTDOWN_ARG, flushFinal, fields.add(flushOpts(false))
+                SHUTDOWN_ARG, flushFinal, fields.add("flushFlag", flushOpts(false))
```

Comment:
I can see `flushArg` and `shutdownArg` in the generated result. Can you show me a sample pipeline that prints `flushFinalFlag` and `flushFlag`?

Reply:
This can be triggered by filter plugins that expose `flush` methods, like `aggregate` and `throttle`.
For example
```
input {
	stdin {}
}

filter {
	throttle {
		key => "my_key"
	}
}

output {
	stdout {}
}
```
From:
```java
public final class CompiledDataset1 extends org.logstash.config.ir.compiler.BaseDataset
    implements org.logstash.config.ir.compiler.Dataset {
  private final java.util.ArrayList field0;
  private final org.logstash.config.ir.compiler.FilterDelegatorExt field1;
  private final org.jruby.RubyHash field2;
  private final org.jruby.RubyHash field3;

  public CompiledDataset1(java.util.Map arguments) {
    field0 = ((java.util.ArrayList) arguments.get("field0"));
    field1 = ((org.logstash.config.ir.compiler.FilterDelegatorExt) arguments.get("field1"));
    field2 = ((org.jruby.RubyHash) arguments.get("field2"));
    field3 = ((org.jruby.RubyHash) arguments.get("field3"));
  }

  public java.util.Collection compute(
      org.jruby.RubyArray batchArg, boolean flushArg, boolean shutdownArg) {
    if (this.isDone()) {
      return field0;
    }
    org.apache.logging.log4j.ThreadContext.put("plugin.id", field1.getId().toString());
    field0.addAll(field1.multiFilter(batchArg));
    if (flushArg) {
      field0.addAll(field1.flush((shutdownArg ? field2 : field3)));
    }
    org.apache.logging.log4j.ThreadContext.remove("plugin.id");
    this.setDone();
    return field0;
  }

  public void clear() {
    if (this.isDone()) {
      field0.clear();
      this.clearDone();
    }
  }
}
```

Become now:
```java
public final class CompiledDataset1 extends org.logstash.config.ir.compiler.BaseDataset
    implements org.logstash.config.ir.compiler.Dataset {
  private final java.util.ArrayList outputBufferField;
  private final org.logstash.config.ir.compiler.FilterDelegatorExt pluginField;
  private final org.jruby.RubyHash shutdownFlushOptionsField;
  private final org.jruby.RubyHash flushOptionsField;

  public CompiledDataset1(java.util.Map arguments) {
    outputBufferField = ((java.util.ArrayList) arguments.get("outputBufferField"));
    pluginField =
        ((org.logstash.config.ir.compiler.FilterDelegatorExt) arguments.get("pluginField"));
    shutdownFlushOptionsField = ((org.jruby.RubyHash) arguments.get("shutdownFlushOptionsField"));
    flushOptionsField = ((org.jruby.RubyHash) arguments.get("flushOptionsField"));
  }

  public java.util.Collection compute(
      org.jruby.RubyArray batchArg, boolean flushArg, boolean shutdownArg) {
    if (this.isDone()) {
      return outputBufferField;
    }
    org.apache.logging.log4j.ThreadContext.put("plugin.id", pluginField.getId().toString());
    outputBufferField.addAll(pluginField.multiFilter(batchArg));
    if (flushArg) {
      outputBufferField.addAll(
          pluginField.flush((shutdownArg ? shutdownFlushOptionsField : flushOptionsField)));
    }
    org.apache.logging.log4j.ThreadContext.remove("plugin.id");
    this.setDone();
    return outputBufferField;
  }

  public void clear() {
    if (this.isDone()) {
      outputBufferField.clear();
      this.clearDone();
    }
  }
}
```

### Note
Thanks to your observation I renamed `flushFlag` - >`flushOptionsField` and `flushFlag ` -> `shutdownFlushOptionsField`

Reply:
The output, for above pipeline
From:
```java
public final class CompiledDataset2 extends org.logstash.config.ir.compiler.BaseDataset
    implements org.logstash.config.ir.compiler.Dataset {
  private final org.logstash.config.ir.compiler.OutputDelegatorExt field0;

  public CompiledDataset2(java.util.Map arguments) {
    field0 = ((org.logstash.config.ir.compiler.OutputDelegatorExt) arguments.get("field0"));
  }

  public java.util.Collection compute(
      org.jruby.RubyArray batchArg, boolean flushArg, boolean shutdownArg) {
    org.apache.logging.log4j.ThreadContext.put("plugin.id", field0.getId().toString());
    field0.multiReceive(batchArg);
    org.apache.logging.log4j.ThreadContext.remove("plugin.id");
    return null;
  }

  public void clear() {}
}
```

to:
```java
public final class CompiledDataset2 extends org.logstash.config.ir.compiler.BaseDataset
    implements org.logstash.config.ir.compiler.Dataset {
  private final org.logstash.config.ir.compiler.OutputDelegatorExt outputDelegatorField;

  public CompiledDataset2(java.util.Map arguments) {
    outputDelegatorField =
        ((org.logstash.config.ir.compiler.OutputDelegatorExt)
            arguments.get("outputDelegatorField"));
  }

  public java.util.Collection compute(
      org.jruby.RubyArray batchArg, boolean flushArg, boolean shutdownArg) {
    org.apache.logging.log4j.ThreadContext.put(
        "plugin.id", outputDelegatorField.getId().toString());
    outputDelegatorField.multiReceive(batchArg);
    org.apache.logging.log4j.ThreadContext.remove("plugin.id");
    return null;
  }

  public void clear() {}
}
```